### DISCUSSION TITLE
Source Directivity

### DISCUSSION THREAD

**Discussion by Clarity-Acoustical** [Ifsttar/I-Simpa#282] [Q&A]
Hello!

I'm looking for more information on setting up source directivity.

For the simulation I'm working on, I need a source with (roughly) a 180 degree horizontal spread, and a ~30-50 degree vertical spread.  I did see another post (https://groups.google.com/g/i-simpa-user-forum/c/hNgM9BfOT8M/m/aXTo-k9SBQAJ) which indicated I-Simpa uses the EASE format for user-defined directivities.  I found some manufacturer .gll files (i.e., https://vue-audiotechnik.com/downloads-glls/#, and https://jblpro.com/en/support_downloads/download_types/easedatum), however, when I use these files (I have to change the extension from .gll to .txt for I-Simpa to recognize them), the simulation is unable to be completed.

Can anyone provide some information?

Thanks,
Luke

**Comment by nicolas-f** [comment]
Hi
You have to generate a file with the same format as written in the files located in the I-Simpa installation folder:
Directivities\speaker-test1.txt

This format is defined here:
http://www.olsonsound.com/PDF/EASE%204.0%20Loudspeaker%20Device%20File%20Formats%20v4.04e.pdf

**Reply by Clarity-Acoustical** [reply]
I see, that makes sense, thank you.

For my simulation, it's not critical that I'm modeling a specific speaker; but I do need the source directivity to be close to the specified angles.

Is there a database of speaker directivity files that are in this format?  Or, alternatively, is there a straightforward way to generate a file like this for a given set of vertical and horizontal angles?

Thanks again!

**Reply by nicolas-f** [reply]
At the time of implementation there was no open format for directionality, so unfortunately there is no compatible database at the moment. With the addition of the OpenDaff driver in the future this might change. http://www.opendaff.org/

**Reply by Clarity-Acoustical** [reply]
Thanks for the help so far.  I've been engaging in the laborious project of creating my own directivity file to meet my needs.  

But I have a question on how I-Simpa reads the directivity files.  According to the standard you linked, "A positive value represents attenuation in dB relative to the on-axis value, and a negative value indicates the gain in dB relative to the on-axis value."  However, in the included directivites (speaker-test1-3), as the numbers get further off-axis, they get more negative.  So does I-Simpa read a negative value as attenuation relative to on-axis, or does it read a positive value as attenuation relative to on-axis?

Thank you,
Luke

![image](https://user-images.githubusercontent.com/86608807/124028427-56558b00-d9ca-11eb-8b90-e8060c710fca.png)

**Reply by toasty-ghost** [reply]
Hi Luke,

You may have already finished your manual specification by now, but if not you could consider using VACS by R&D-Team to generate frequency response curves based on an input equation, and merge these into horizontal and vertical polar maps within the same app. It can export EASE-compatible or formatted balloon data too.

Alternatively, you could combine VACS with their Akabak boundary element method modelling software. You could model the entire speaker if you like, but it also offers a 'pressure box' element where you can set the vertical and horizontal directivity based on simple shaped pistons, or a line of point sources of a designated length and arc. There's an observation type for a 'shell' which can be set to a sphere for the creation of near-field balloon (NFB) data too, and the calculation is super fast without the addition of boundaries.

Here's a quick example:
![image](https://user-images.githubusercontent.com/46981655/147677431-c3aa23e7-0cfc-450a-8fa0-6c45dd3f0dd5.png)
![image](https://user-images.githubusercontent.com/46981655/147678939-41e85799-9f64-487e-bd56-57980e8c0b48.png)
![image](https://user-images.githubusercontent.com/46981655/147678591-d371255f-98c5-4fab-b3d1-cbb0e2bb622f.png)
![image](https://user-images.githubusercontent.com/46981655/147678876-810cf364-45cc-4660-966d-d19894d99645.png)

You'd need to experiment to get what you need, but with a bit of work you could likely make a model of a circular low-frequency driver and waveguide using the built-in tools, and adjust them to get various directivity patterns you'd need. There's a few examples that come with it which you could modify too. Akabak's data can be opened directly in VACS, where you can export in the EASE-compatible format. Both bits of software are free to use (with minor restrictions) and have great help documentation.
http://www.randteam.de/Index.html

Hope that helps!

**Reply by ldaniel-jmz** [reply]
> Hi You have to generate a file with the same format as written in the files located in the I-Simpa installation folder: Directivities\speaker-test1.txt
> 
> This format is defined here: http://www.olsonsound.com/PDF/EASE%204.0%20Loudspeaker%20Device%20File%20Formats%20v4.04e.pdf

Haven´t been able to generate the directivity EASE ASCII file. I have all the data needed and I'm using the Excel template provided by EASE "EASE40_SpeakerModel_Data.xlt", but when I click in the button "Save data as EASE ASCII file" it shows an error. I think this file is too old.

Does someone have a newer working template file to generate the ASCII file? Or is there another way to generate this files?

I've done a lot of research and haven't been able to find a solution.

**Reply by nicolas-f** [reply]
Hi,

In which format is your original directivity file?

**Reply by ldaniel-jmz** [reply]
I just have the data in a spreadsheet.

**Reply by nicolas-f** [reply]
As you can see here:
https://github.com/Universite-Gustave-Eiffel/I-Simpa/blob/master/currentRelease/Directivities/speaker-test2.txt
and here (the directivity file reader):
https://github.com/Universite-Gustave-Eiffel/I-Simpa/blob/v1.3.4/src/lib_interface/input_output/directivity/directivityParser.cpp#L69

I-Simpa expect the following line:
`"Frequency",40,"Hz" `
It reads the frequency number

then for each line it will read the phi number and all attenuation for each theta values (37 * 5° ) 0-185

So you should write a script in order to generate this file from your spreadsheet. And check the orientation with a simulation in SPPS.

**Reply by ldaniel-jmz** [reply]
I don´t have the knowledge to generate that script, that's why I was trying to use the Excel template that generates the ASCII .xhn directivity files provided by amfg, as you mentioned in a previous discussion I-Simpa uses tabular text format directivity as specified by EASE.

 This template is of free-use, and you can download it [here](https://www.afmg.eu/en/ms-excel-templates-ease-4x-data).

**Reply by fredmzlp** [reply]
Hi all,
I am also trying to get a custom directivity pattern to I-Simpa. 
By looking at the directivity files on GitHub, and at the format definition pdf (http://www.olsonsound.com/PDF/EASE%204.0%20Loudspeaker%20Device%20File%20Formats%20v4.04e.pdf) a couple questions arise.

The pdf EASE definition of azimuth and elevations quite odd (?). And the way it recommends arranging the data is with theta in rows, going from 0 to 180, and with phi in columns, going from 0 to 355. 
Checking the directivity speaker-test3.txt, I notice that rows go from 0 to 355, while columns go from 0 to 180. What is the convention for direction angles applied here? 
I would assume that rows from 0 to 355 are azimuth angles (phi), while columns are elevation angles (theta). According to the documentation "Modelling of physical phenomena in SPPS" figure "Elementary geometry for a sound source" this would mean angles (0,0) as direction of z axis (i.e direction pointing to directly above?)
Though, defining an omnidirectional source, further below in the documentation, polar plot of "Distribution of emission angles 𝜃
for an omnidirectional source" seems not compatible to theta defined from 0 to 180, and omnidirectional radiation as a "sphere". I probably am missing something regarding this plot...?

Find figures attached for convenience.

And as a final observation, speaker-test1.txt and speaker-test2.txt files do not have the header columns 0 to 180 angles, as in speaker-test3.txt. Are this not strictly necessary for I-Simpa?

Hoping I was clear enough expressing my questions.
Any contribution is much appreciated

Thanks, Fred

<img width="714" alt="Screen Shot 2023-04-11 at 4 50 31 pm" src="https://user-images.githubusercontent.com/128641443/231222655-7ac531a6-336c-401e-a5aa-e8f04d8eaa9c.png">
<img width="519" alt="Screen Shot 2023-04-11 at 4 50 04 pm" src="https://user-images.githubusercontent.com/128641443/231222666-7a4fc2f8-78d8-4836-85b5-d9c65a0a6107.png">
<img width="521" alt="Screen Shot 2023-04-11 at 4 49 54 pm" src="https://user-images.githubusercontent.com/128641443/231222674-605313cb-9da4-4bd0-91dd-050cc49635d1.png">

**Reply by ArmandDuverger** [reply]
Hi everyone,
I'd like to import speaker data from a GLL file to I-Simpa and I have the exact same questions. Did you find answers?
Thanks by advance!

**Reply by ArmandDuverger** [reply]
I wrote a [import script](https://github.com/ArmandDuverger/gll_to_isimpa) from GLL to I-Simpa txt (EASE ASCII 4) which seems to work for me.
